mobile/ios/ios-proxy-services.ts

```diff
@@ -220,26 +221,108 @@ export class InstallationProxyClient {
 		private $injector: IInjector) { }
 
-	postNotification(notificationName: string) {
-		this.plistService = this.$injector.resolve(iOSCore.PlistService, {service:  this.device.startService(MobileServices.NOTIFICATION_PROXY), format: iOSCore.CoreTypes.kCFPropertyListBinaryFormat_v1_0});
-
+	public postNotification(notificationName: string) {
+        this.openSocket();
 		var result = this.plistService.sendMessage({
 			"Command": "PostNotification",
 			"Name": notificationName,
 			"ClientOptions": ""
 		});
-
 		return result;
-	}
+    }
+
+    public addObserver(name: string, callback: (name: string) => void) {
+        this.openSocket();
+
+        var result = this.plistService.sendMessage({
+            "Command": "ObserveNotification",
+            "Name": name
+        });
+
+        var array = this.observers[name];
+        if (!array) {
+            array = new Array();
+            this.observers[name] = array;
+        }
+        array.push(callback);
+
+        return result;
+    }
+
+    public removeObserver(name: string, callback: (name: string) => void) {
+        var array = this.observers[name];
+        if (array) {
+            var index = array.indexOf(callback);
+            if (index >= 0 && index < array.length) {
+                array.splice(index, 1);
+            }
+        }
+    }
+
+    private openSocket() {
+        if (!this.plistService) {
+            this.plistService = this.$injector.resolve(iOSCore.PlistService, { service: this.device.startService(MobileServices.NOTIFICATION_PROXY), format: iOSCore.CoreTypes.kCFPropertyListBinaryFormat_v1_0 });
+            this.plistService.readSystemLog(this.handleData.bind(this));
+        }
+    }
+
+    private handleData(data: NodeBuffer): void {
+        this.buffer += data.toString();
+
+        var start = this.buffer.indexOf("<plist");
+        var end = this.buffer.indexOf("</plist>") + 8;
+
+        while (start >= 0 && end >= 0) {
+            var plist = this.buffer.substr(start, end);
+            this.buffer = this.buffer.substr(end);
+
+            try {
+                plistlib.loadString(plist, (err: any, plist: any) => {
+                    if (err) {
+                    } else if (plist) {
+                        this.handlePlistNotification(plist);
+                    } else {
+                    }
+                });
+            }
+            catch (e) {
+            }
+
+            var start = this.buffer.indexOf("<plist");
+            var end = this.buffer.indexOf("</plist>") + 8;
+        }
+    }
 
 	public closeSocket() {
 		this.plistService.close();
-	}
+    }
+
+    private handlePlistNotification(plist: any) {
+        if (plist.type !== "dict")
+            return;
+        var value = plist.value;
+        if (!value)
+            return;
+        var command = value["Command"];
+        var name = value["Name"]
+        if (command.type !== "string" || command.value !== "RelayNotification" || name.type !== "string")
```

Comment:
There is (perhaps very small) probability that we may get plist with another command.